lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp

```diff
@@ -0,0 +1,69 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Dialect/TTNN/Analysis/L1InterleavedPolicy.h"
+#include "ttmlir/Dialect/TT/IR/TTOpsTypes.h"
+#include "ttmlir/Scheduler/Scheduler.h"
+#include <llvm/Support/raw_ostream.h>
+
+namespace mlir::tt::ttnn {
+
+void L1InterleavedPolicy::run() {
+  rootOp->walk([&](func::FuncOp func) {
+    mlir::tt::scheduler::Scheduler scheduler(&func);
+    llvm::SmallVector<mlir::Operation *> scheduleableOps;
+    Operation *currentOp = nullptr;
+    llvm::DenseMap<Operation *, tt::LayoutAttr> selectedOpLayout;
+
+    // TODO(fbajraktari): Algo
+    //
+    l1ChainConfigs->push_back(L1ChainConfig());
+    while (scheduler.hasUnscheduledOps()) {
+      scheduleableOps = scheduler.getScheduleableOps();
+      currentOp = scheduleableOps[0];
+
+      // Schedule currentOp.
+      //
+      scheduler.scheduleOp(currentOp);
+
+      // Check if currentOp is valid l1 interleaved op.
+      //
+      if (legalLayouts.lookup(currentOp).size() > 0) {
+        selectedOpLayout[currentOp] = legalLayouts.lookup(currentOp).front();
+
+        // Add currentOp to l1 chain config.
+        //
+        OpL1MemSpec shardSpec;
+        shardSpec.op = currentOp;
+
+        // Hardcoded tensor split factor for now, until pipeline OP
+        // support is added.
+        //
+        shardSpec.tensorSplitFactor = 1;
+        l1ChainConfigs->back().addOpL1MemSpec(std::move(shardSpec));
```

Comment:
If i understand this correctly, we are adding everything to a single chain. So when we have an op that does not have L1 interleaved as a legal layout, what happens? Do we have a testcase for this?

Reply:
In the V0 implementation, we insert all ops with l1 interleaved memory layout into a single "chain" even though they may be disconnected. For an op that doesn't support l1 interleaved, OpConfigAnalysis will pick default layout which is DRAM.

We don't have test for this since LegalGridAnalysis constructs l1 interleaved layout for every op. Is there any op that currently doesn't support l1 interleaved?

Reply:
Ok, let's leave as is for V0 implementation and then make sure that for V1 we don't have disconnected chains. Leave a TODO comment for this somewhere in the code. 

>We don't have test for this since LegalGridAnalysis constructs l1 interleaved layout for every op. Is there any op that currently doesn't support l1 interleaved?
We can test this with overriding output layout to dram, in that case LegalGridAnalysis will not set l1 interleaved as a possibility.